### DISCUSSION TITLE
Unclear how to properly get type info in VS Code

### DISCUSSION THREAD

**Discussion by samestep** [tree-sitter/tree-sitter#3853] [Q&A]
### Problem

I installed the Tree-sitter CLI using Cargo. After running `tree-sitter init` as instructed in the Tree-sitter docs on [Creating parsers](https://tree-sitter.github.io/tree-sitter/creating-parsers), when I open up `grammar.js` in VS Code, I get the following inline error messages:

- Cannot find type definition file for 'tree-sitter-cli/dsl'.
- Cannot find name 'grammar'.

I was able to get rid of those errors by running `npm install` which is not mentioned in the docs, but on my machine (I use [nvm](https://github.com/nvm-sh/nvm) and [pyenv](https://github.com/pyenv/pyenv), and currently have Node v18.20.3 and Python v3.12.3 selected as my defaults) it fails with this error:

```
$ npm install        

> tree-sitter-carbon@0.1.0 install
> node-gyp-build

gyp info it worked if it ends with ok
gyp info using node-gyp@10.1.0
gyp info using node@18.20.3 | darwin | arm64
gyp info find Python using Python version 3.12.3 found at "/Users/samueles/.pyenv/versions/3.12.3/bin/python3"

gyp info spawn /Users/samueles/.pyenv/versions/3.12.3/bin/python3
gyp info spawn args [
gyp info spawn args '/Users/samueles/.nvm/versions/node/v18.20.3/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args 'binding.gyp',
gyp info spawn args '-f',
gyp info spawn args 'make',
gyp info spawn args '-I',
gyp info spawn args '/Users/samueles/github/samestep/tree-sitter-carbon/build/config.gypi',
gyp info spawn args '-I',
gyp info spawn args '/Users/samueles/.nvm/versions/node/v18.20.3/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args '-I',
gyp info spawn args '/Users/samueles/Library/Caches/node-gyp/18.20.3/include/node/common.gypi',
gyp info spawn args '-Dlibrary=shared_library',
gyp info spawn args '-Dvisibility=default',
gyp info spawn args '-Dnode_root_dir=/Users/samueles/Library/Caches/node-gyp/18.20.3',
gyp info spawn args '-Dnode_gyp_dir=/Users/samueles/.nvm/versions/node/v18.20.3/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args '-Dnode_lib_file=/Users/samueles/Library/Caches/node-gyp/18.20.3/<(target_arch)/node.lib',
gyp info spawn args '-Dmodule_root_dir=/Users/samueles/github/samestep/tree-sitter-carbon',
gyp info spawn args '-Dnode_engine=v8',
gyp info spawn args '--depth=.',
gyp info spawn args '--no-parallel',
gyp info spawn args '--generator-output',
gyp info spawn args 'build',
gyp info spawn args '-Goutput_dir=.'
gyp info spawn args ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
  TOUCH Release/obj.target/node_modules/node-addon-api/node_addon_api_except.stamp
  CXX(target) Release/obj.target/tree_sitter_carbon_binding/bindings/node/binding.o
In file included from ../bindings/node/binding.cc:1:
../node_modules/node-addon-api/napi.h:14:10: fatal error: 'functional' file not found
   14 | #include <functional>
      |          ^~~~~~~~~~~~
1 error generated.
make: *** [Release/obj.target/tree_sitter_carbon_binding/bindings/node/binding.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack at ChildProcess.<anonymous> (/Users/samueles/.nvm/versions/node/v18.20.3/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:209:23)
gyp ERR! System Darwin 24.0.0
gyp ERR! command "/Users/samueles/.nvm/versions/node/v18.20.3/bin/node" "/Users/samueles/.nvm/versions/node/v18.20.3/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/samueles/github/samestep/tree-sitter-carbon
gyp ERR! node -v v18.20.3
gyp ERR! node-gyp -v v10.1.0
gyp ERR! not ok 
npm error code 1
npm error path /Users/samueles/github/samestep/tree-sitter-carbon
npm error command failed
npm error command sh -c node-gyp-build

npm error A complete log of this run can be found in: /Users/samueles/.npm/_logs/2024-10-30T17_55_06_490Z-debug-0.log
```

### Steps to reproduce

```
mkdir tree-sitter-foo
cd tree-sitter-foo
tree-sitter init
npm install
```

### Expected behavior

I expected the docs page to mention that `npm install` needs to be run in order to get rid of the error messages in VS Code and give docs and type info on hover. I also expected that `npm install` would succeed. If a specific version for Node and/or Python are required, I would expect those to be listed in the docs.

### Tree-sitter version (tree-sitter --version)

tree-sitter 0.24.3

### Operating system/version

macOS Sequoia 15.0.1

**Comment by clason** [comment]
I'm sorry, but this is a VS Code issue and unrelated to tree-sitter.

**Reply by samestep** [reply]
@clason I'm sorry, could you clarify? This is not specific to VS Code; it's about the fact that Tree-sitter expects `node_modules` to exist and yet does not provide any docs for running `npm install` or similar to create `node_modules`. Or am I missing something? In any case, there's definitely a gap in the docs.

**Reply by samestep** [reply]
In particular, the fact that `npm install` fails is very much not related to VS Code.

**Reply by samestep** [reply]
@clason it seems like you simply closed this issue because I had "VS Code" in the title; if the title instead simply said that `npm install` fails after `tree-sitter init`, would you not have closed it so hastily?

**Reply by clason** [reply]
You don't _need_ type information for tree-sitter to work, so this is very much a VS Code (or language server) issue only. `npm install` is normally not required.

And there's a lot of context missing here to help you further.

**Reply by samestep** [reply]
@clason what context is missing? I've provided details on exactly what versions of everything I have installed, and how I installed it.

**Reply by clason** [reply]
Missing are exact, reproducible steps. `tree-sitter init` prompts you for input; what did you enter? Did you follow the steps in the docs _exactly_ (which, I repeat, do _not_ call for `npm install` right after `tree-sitter init`)?

**Reply by samestep** [reply]
See #3854.

**Comment by samestep** [comment]
Closing as a duplicate of #3854.
